### PR TITLE
[perf] feat: support DeepSeek V4 MFU calculation

### PR DESCRIPTION
**Pull request by FoolPlayer**
### What does this PR do?

> Concise overview of the change. Reference related issues/PRs.

### Checklist Before Starting

- Search for relative PRs/issues and link here: ...
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> Validation results (training curves, eval metrics) for changes not covered by CI.

### API and Usage Example

> Show API changes and usage examples if applicable.

### Design & Code Changes

> High-level design description and specific change list.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

File: veomni/utils/count_flops.py [hunk 1]
```diff
@@ -185,6 +186,123 @@ def _estimate_deepseek_v3_flops(self, tokens_sum, batch_seqlens, delta_time):
...
+        # Shared-KV MQA projections plus the grouped low-rank output projection.
+        attn_linear_N = hidden_size * q_lora_rank
+        attn_linear_N += q_lora_rank * num_attention_heads * head_dim
+        attn_linear_N += hidden_size * head_dim
+        attn_linear_N += num_attention_heads * head_dim * o_lora_rank
+        attn_linear_N += o_groups * o_lora_rank * hidden_size
+
+        # Every layer computes router logits, top-k routed experts, and one shared expert.
+        moe_router_N = hidden_size * num_experts
+        moe_expert_N = hidden_size * moe_intermediate_size * (experts_per_token + 1) * 3
```
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply veomni/utils/count_flops.py]: Thanks — agreed.

Fixed in 5f90a391: DeepSeek-V4 FLOPs now reads `n_shared_experts` from config (defaulting to 1 / treating falsy as 0), matching the DeepSeek-V3 path, and added `test_shared_experts_scale_moe_flops`.

File: veomni/utils/count_flops.py [hunk 2]
```diff
@@ -185,6 +186,123 @@ def _estimate_deepseek_v3_flops(self, tokens_sum, batch_seqlens, delta_time):
...
+        if num_csa_layers > 0:
+            csa_compress_rate = self.config.compress_rates["compressed_sparse_attention"]
+            csa_index_score_sum = self._compute_compressed_attention_score_sum(batch_seqlens, csa_compress_rate)
+            csa_attention_score_sum = self._compute_compressed_attention_score_sum(
+                batch_seqlens, csa_compress_rate, topk=self.config.index_topk
+            )
+
+        attention_score_sum = sliding_score_sum * num_hidden_layers
+        attention_score_sum += hca_score_sum * num_hca_layers
+        attention_score_sum += csa_attention_score_sum * num_csa_layers
```
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply veomni/utils/count_flops.py:298]: Thanks for flagging this — I looked into it carefully and **don't think this is a double-count**.

DeepSeek-V4 attention is not "either sliding or CSA/HCA". Modeling documents three mechanisms: **sliding / sliding+CSA / sliding+HCA**. For CSA/HCA layers, the compressor concatenates compressed KV onto the existing sliding-window KV axis, and the attention call still passes `sliding_window`. So pure `sliding_attention` only means "no compressor", not "only these layers have sliding scores".

Changing to `sliding_score_sum * num_sliding_layers` would undercount CSA/HCA layers (especially configs with zero pure-sliding layers).

Left the multiplication by `num_hidden_layers` as-is, and added an explicit comment in 5f90a391 to prevent this misread later.

File: tests/utils/test_count_flops.py [hunk 3]
```diff
@@ -148,3 +166,33 @@ def test_sliding_attention_reduces_quadratic_flops(self, gpt_oss_config):
...
         assert full_flops > mixed_flops
+
+
+class TestDeepseekV4Flops:
+    pytestmark = pytest.mark.usefixtures("mock_device_flops")
+
+    def test_numerical(self, deepseek_v4_counter):
+        flops, promised_flops = deepseek_v4_counter.estimate_flops([12, 5], delta_time=1.0)
+
+        assert flops == pytest.approx(0.000264658944, rel=1e-9)
```
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply tests/utils/test_count_flops.py:177]: Keeping `0.000264658944`.

That expected value matches the intended model: sliding-window scores on every decoder layer (including CSA/HCA). The suggested `0.000264373248` assumes sliding is only counted on pure sliding layers, which undercounts V4 as implemented — see the sibling thread on `sliding_score_sum * num_hidden_layers`.

File: veomni/utils/count_flops.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply veomni/utils/count_flops.py]: Yes, should read n_shared_experts from config

File: veomni/utils/count_flops.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply veomni/utils/count_flops.py:298]: CSA/HCA still keep sliding-window KV path

File: tests/utils/test_count_flops.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply tests/utils/test_count_flops.py:177]: Architecture does not support the rewrite

**Review by Luosuu — APPROVED** [review]
